org.eclipse.xtext/src/org/eclipse/xtext/formatting2/internal/HiddenRegionReplacer.java

```diff
@@ -56,26 +68,52 @@ protected void applyHiddenRegionFormatting(List<ITextReplacer> replacers) {
 		} else if (dec != null) {
 			separator.setIndentationDecrease(dec);
 		}
+	}
 
-		// 2. apply NewLine and space configuration to the first whitespace region that follows or contains a linewrap.  
+	/**
+	 * Apply NewLine and space configuration to the first whitespace region that follows or contains a linewrap.
+	 * @since 2.15 
+	 */
+	protected void applySeparatorConfiguration(IHiddenRegionFormatting separator) {
 		separator.setNewLinesDefault(formatting.getNewLineDefault());
 		separator.setNewLinesMax(formatting.getNewLineMax());
 		separator.setNewLinesMin(formatting.getNewLineMin());
 		separator.setSpace(formatting.getSpace());
+	}
 
-		// 3. apply the 'priority' configuration for all and set a default formatting
-		for (ITextReplacer replacer : replacers)
+	/**
+	 * @since 2.15
+	 */
+	protected void applyPriorityAndDefaultFormatting(List<ITextReplacer> replacers, IHiddenRegionFormatting separator) {
+		for (int i = 0; i < replacers.size(); i++) {
+			ITextReplacer replacer = replacers.get(i);
 			if (replacer instanceof WhitespaceReplacer) {
 				IHiddenRegionFormatting formatting2 = ((WhitespaceReplacer) replacer).getFormatting();
 				formatting2.setPriority(formatting.getPriority());
 				if (formatting2 != separator) {
-					formatting2.setSpace(replacer.getRegion().getOffset() > 0 ? " " : "");
+					ITextReplacer previous = (i == 0) ? null : replacers.get(i - 1);
+					ITextReplacer next = (i + 1 >= replacers.size()) ? null : replacers.get(i + 1);
+					// Don't add single spaces after comments
+					if (previous == null || !(previous instanceof CommentReplacer)) {
+						// Don't add spaces before ML comments because in the next step
+						// a newline will be added in front of ML comments.
+						// You don't want to end up with a ML comment on a newline that
+						// starts with a space
```

Comment:
Isn't this comment true for other tokens, too?

Reply:
Yes, but we're already in this case we know we're dealing with whitespace (see the comment on applyHiddenRegionFormatting). The only tokens we can get here are SL/ML comment or WS.